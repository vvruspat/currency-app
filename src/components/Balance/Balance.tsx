type BalanceProps = {
  balance: number;
  currency: string;
  locale?: string;
} & React.DetailedHTMLProps<
  React.HTMLAttributes<HTMLDivElement>,
  HTMLDivElement
>;

export const Balance = ({
  balance,
  currency,
  locale = "en-US",
  ...divProps
}: BalanceProps) => {
  return (
    <div {...divProps}>
      Balance:{" "}
      {new Intl.NumberFormat(locale, {
        style: "currency",
        currency: currency,
      }).format(balance)}
    </div>
  );
};
