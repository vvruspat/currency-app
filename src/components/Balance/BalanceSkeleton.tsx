type BalanceSkeletronProps = React.DetailedHTMLProps<
  React.HTMLAttributes<HTMLDivElement>,
  HTMLDivElement
>;

export const BalanceSkeleton = ({ ...divProps }: BalanceSkeletronProps) => {
  return (
    <div {...divProps}>
      Balance: <div className="skeleton"></div>
    </div>
  );
};
