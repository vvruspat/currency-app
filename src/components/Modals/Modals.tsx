import React, { useCallback, useState } from "react";
import {
  SelectModal,
  SelectModalProps,
} from "../uikit/Select/SelectModal/SelectModal";

export const ModalsContext = React.createContext<{
  setModal: (modal: ModalsProps | null) => void;
}>({
  setModal: (modal: ModalsProps | null) => {},
});

export enum MODALS {
  SELECT_MODAL = "select-modal",
  ERROR_MODAL = "error-modal",
  EXCHANGE_SUCCESS_MODAL = "exchange-success-modal",
}

// TODO: render and start modals from the component where it need

type ModalsProps = {
  modal: MODALS.SELECT_MODAL;
  props: SelectModalProps;
};
// | { modal: MODALS.ERROR_MODAL; props: /*ErrorModalProps*/ {} }
// | {
//     modal: MODALS.EXCHANGE_SUCCESS_MODAL;
//     props: /*ExchangeSuccessModalProps*/ {};
//   };

type ModalsRootProps = {};

export const ModalsRoot: React.FC<ModalsRootProps> = ({ children }) => {
  const [currentModal, setCurrentModal] = useState<ModalsProps | null>(null);

  const setModal = useCallback((modal: ModalsProps | null) => {
    setCurrentModal(modal);
  }, []);

  return (
    <ModalsContext.Provider value={{ setModal }}>
      {children}
      {/* <ErrorModal
        {...currentModalProps}
        show={currentModal === MODALS.ERROR_MODAL}
      /> */}
      <SelectModal
        {...currentModal?.props}
        show={currentModal?.modal === MODALS.SELECT_MODAL}
      />
      {/* <ExchangeSuccessModal
        {...currentModalProps}
        show={currentModal === MODALS.EXCHANGE_SUCCESS_MODAL}
      /> */}
    </ModalsContext.Provider>
  );
};
