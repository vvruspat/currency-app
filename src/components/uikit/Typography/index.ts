export * from "./Title";
export * from "./Caption";
