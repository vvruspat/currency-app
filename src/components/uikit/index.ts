export * from "./Header";
export * from "./Panel";
export * from "./Group";
export * from "./Button";
export * from "./Input";
export * from "./Modal";
export * from "./IconButton";
export * from "./Typography";
