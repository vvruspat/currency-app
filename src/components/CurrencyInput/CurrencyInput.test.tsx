import { render, screen } from "@testing-library/react";
import { CurrencyDirection, CurrencyInput } from ".";
import { Balance } from "../Balance";
import { MODALS } from "../Modals/Modals";
import { Caption, SelectorOption } from "../uikit";

const options = [
  {
    key: "USD",
    value: {
      icon: <div data-testid="test1-icon">Test1 icon element</div>,
      description: <Caption>USD</Caption>,
      content: <Balance currency="USD" balance={5676.56} />,
    },
  },
  {
    key: "EUR",
    value: {
      icon: <div data-testid="test2-icon">Test2 icon element</div>,
      description: <Caption>EUR</Caption>,
      content: <Balance currency="EUR" balance={1276.76} />,
    },
  },
];

test("renders CurrencyInput noerror in", () => {
  render(
    <CurrencyInput
      data-testid="test-currency-input"
      options={options}
      balance={6655.67}
      amount={665.76}
      error={undefined}
      onCurrencyAmountChange={(amount: number) => {}}
      onCurrencyChange={(option: SelectorOption) => {}}
      direction={CurrencyDirection.IN}
      modalNav={MODALS.TEST_MODAL}
    />,
    {}
  );
  const CurrencyInputElement = screen.getByTestId("test-currency-input");

  expect(CurrencyInputElement).toBeInTheDocument();
  expect(CurrencyInputElement).toMatchSnapshot();
});

test("renders CurrencyInput noerror out", () => {
  render(
    <CurrencyInput
      data-testid="test-currency-input"
      options={options}
      balance={1655.67}
      amount={65.46}
      error={undefined}
      onCurrencyAmountChange={(amount: number) => {}}
      onCurrencyChange={(option: SelectorOption) => {}}
      direction={CurrencyDirection.OUT}
      modalNav={MODALS.TEST_MODAL}
    />,
    {}
  );
  const CurrencyInputElement = screen.getByTestId("test-currency-input");

  expect(CurrencyInputElement).toBeInTheDocument();
  expect(CurrencyInputElement).toMatchSnapshot();
});

test("renders CurrencyInput error out", () => {
  render(
    <CurrencyInput
      data-testid="test-currency-input"
      options={options}
      balance={1655.67}
      amount={4565.46}
      error={"balance exceed"}
      onCurrencyAmountChange={(amount: number) => {}}
      onCurrencyChange={(option: SelectorOption) => {}}
      direction={CurrencyDirection.OUT}
      modalNav={MODALS.TEST_MODAL}
    />,
    {}
  );
  const CurrencyInputElement = screen.getByTestId("test-currency-input");

  expect(CurrencyInputElement).toBeInTheDocument();
  expect(CurrencyInputElement).toMatchSnapshot();
});
