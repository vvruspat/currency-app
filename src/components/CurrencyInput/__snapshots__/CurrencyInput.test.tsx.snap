// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CurrencyInput error out 1`] = `
<div
  class="currency-input"
  data-testid="test-currency-input"
>
  <div
    class="currency-input-controls"
  >
    <div
      data-testid="select-test"
    >
      <button
        class="icon-button icon-right"
        data-testid="select-buton"
      >
        <div
          class="icon-button-content"
        >
          <div
            class="icon-button-icon"
          >
            <svg>
              Down16Icon.svg
            </svg>
          </div>
          <div
            class="icon-button-text"
          >
            <div>
              Balance:
               
              $5,676.56
            </div>
          </div>
        </div>
      </button>
      <div
        class="modal-overlay modal-overlay-hide"
      />
      <div
        class="modal"
        data-testid="select-modal"
      >
        <header
          class="modal-header"
        >
          <div
            class="modal-header-content"
          >
            <div
              class="caption"
            >
              <input
                class="input input-align-left select-modal-filter-input"
                data-testid="select-modal-filter-input"
                placeholder="Search"
                value=""
              />
            </div>
          </div>
          <button
            class="icon-button modal-header-close-button"
          >
            <div
              class="icon-button-content"
            >
              <div
                class="icon-button-icon"
              >
                <svg>
                  Cross16Icon.svg
                </svg>
              </div>
              <div
                class="icon-button-text"
              >
                Close
              </div>
            </div>
          </button>
        </header>
        <section
          class="modal-content"
        >
          <div
            style="position: relative;"
          >
            <div
              class="select-option"
            >
              <div
                class="select-option-icon"
              >
                <div
                  data-testid="test1-icon"
                >
                  Test1 icon element
                </div>
              </div>
              <div
                class="select-option-content"
              >
                <div
                  class="select-option-content-data"
                >
                  <div>
                    Balance:
                     
                    $5,676.56
                  </div>
                </div>
                <div
                  class="select-option-content-description"
                >
                  <div
                    class="caption"
                  >
                    USD
                  </div>
                </div>
              </div>
            </div>
            <div
              class="select-option"
            >
              <div
                class="select-option-icon"
              >
                <div
                  data-testid="test2-icon"
                >
                  Test2 icon element
                </div>
              </div>
              <div
                class="select-option-content"
              >
                <div
                  class="select-option-content-data"
                >
                  <div>
                    Balance:
                     
                    €1,276.76
                  </div>
                </div>
                <div
                  class="select-option-content-description"
                >
                  <div
                    class="caption"
                  >
                    EUR
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <input
      class="amount-input"
      inputmode="decimal"
      type="text"
      value="-4,565.46"
    />
  </div>
  <div
    class="currency-input-info"
  >
    <div
      class="currency-input-balance"
    >
      Balance:
       
      $1,655.67
    </div>
    <div
      class="currency-input-error"
    >
      balance exceed
    </div>
  </div>
</div>
`;

exports[`renders CurrencyInput noerror in 1`] = `
<div
  class="currency-input"
  data-testid="test-currency-input"
>
  <div
    class="currency-input-controls"
  >
    <div
      data-testid="select-test"
    >
      <button
        class="icon-button icon-right"
        data-testid="select-buton"
      >
        <div
          class="icon-button-content"
        >
          <div
            class="icon-button-icon"
          >
            <svg>
              Down16Icon.svg
            </svg>
          </div>
          <div
            class="icon-button-text"
          >
            <div>
              Balance:
               
              $5,676.56
            </div>
          </div>
        </div>
      </button>
      <div
        class="modal-overlay modal-overlay-hide"
      />
      <div
        class="modal"
        data-testid="select-modal"
      >
        <header
          class="modal-header"
        >
          <div
            class="modal-header-content"
          >
            <div
              class="caption"
            >
              <input
                class="input input-align-left select-modal-filter-input"
                data-testid="select-modal-filter-input"
                placeholder="Search"
                value=""
              />
            </div>
          </div>
          <button
            class="icon-button modal-header-close-button"
          >
            <div
              class="icon-button-content"
            >
              <div
                class="icon-button-icon"
              >
                <svg>
                  Cross16Icon.svg
                </svg>
              </div>
              <div
                class="icon-button-text"
              >
                Close
              </div>
            </div>
          </button>
        </header>
        <section
          class="modal-content"
        >
          <div
            style="position: relative;"
          >
            <div
              class="select-option"
            >
              <div
                class="select-option-icon"
              >
                <div
                  data-testid="test1-icon"
                >
                  Test1 icon element
                </div>
              </div>
              <div
                class="select-option-content"
              >
                <div
                  class="select-option-content-data"
                >
                  <div>
                    Balance:
                     
                    $5,676.56
                  </div>
                </div>
                <div
                  class="select-option-content-description"
                >
                  <div
                    class="caption"
                  >
                    USD
                  </div>
                </div>
              </div>
            </div>
            <div
              class="select-option"
            >
              <div
                class="select-option-icon"
              >
                <div
                  data-testid="test2-icon"
                >
                  Test2 icon element
                </div>
              </div>
              <div
                class="select-option-content"
              >
                <div
                  class="select-option-content-data"
                >
                  <div>
                    Balance:
                     
                    €1,276.76
                  </div>
                </div>
                <div
                  class="select-option-content-description"
                >
                  <div
                    class="caption"
                  >
                    EUR
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <input
      class="amount-input"
      inputmode="decimal"
      type="text"
      value="665.76"
    />
  </div>
  <div
    class="currency-input-info"
  >
    <div
      class="currency-input-balance"
    >
      Balance:
       
      $6,655.67
    </div>
  </div>
</div>
`;

exports[`renders CurrencyInput noerror out 1`] = `
<div
  class="currency-input"
  data-testid="test-currency-input"
>
  <div
    class="currency-input-controls"
  >
    <div
      data-testid="select-test"
    >
      <button
        class="icon-button icon-right"
        data-testid="select-buton"
      >
        <div
          class="icon-button-content"
        >
          <div
            class="icon-button-icon"
          >
            <svg>
              Down16Icon.svg
            </svg>
          </div>
          <div
            class="icon-button-text"
          >
            <div>
              Balance:
               
              $5,676.56
            </div>
          </div>
        </div>
      </button>
      <div
        class="modal-overlay modal-overlay-hide"
      />
      <div
        class="modal"
        data-testid="select-modal"
      >
        <header
          class="modal-header"
        >
          <div
            class="modal-header-content"
          >
            <div
              class="caption"
            >
              <input
                class="input input-align-left select-modal-filter-input"
                data-testid="select-modal-filter-input"
                placeholder="Search"
                value=""
              />
            </div>
          </div>
          <button
            class="icon-button modal-header-close-button"
          >
            <div
              class="icon-button-content"
            >
              <div
                class="icon-button-icon"
              >
                <svg>
                  Cross16Icon.svg
                </svg>
              </div>
              <div
                class="icon-button-text"
              >
                Close
              </div>
            </div>
          </button>
        </header>
        <section
          class="modal-content"
        >
          <div
            style="position: relative;"
          >
            <div
              class="select-option"
            >
              <div
                class="select-option-icon"
              >
                <div
                  data-testid="test1-icon"
                >
                  Test1 icon element
                </div>
              </div>
              <div
                class="select-option-content"
              >
                <div
                  class="select-option-content-data"
                >
                  <div>
                    Balance:
                     
                    $5,676.56
                  </div>
                </div>
                <div
                  class="select-option-content-description"
                >
                  <div
                    class="caption"
                  >
                    USD
                  </div>
                </div>
              </div>
            </div>
            <div
              class="select-option"
            >
              <div
                class="select-option-icon"
              >
                <div
                  data-testid="test2-icon"
                >
                  Test2 icon element
                </div>
              </div>
              <div
                class="select-option-content"
              >
                <div
                  class="select-option-content-data"
                >
                  <div>
                    Balance:
                     
                    €1,276.76
                  </div>
                </div>
                <div
                  class="select-option-content-description"
                >
                  <div
                    class="caption"
                  >
                    EUR
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <input
      class="amount-input"
      inputmode="decimal"
      type="text"
      value="-65.46"
    />
  </div>
  <div
    class="currency-input-info"
  >
    <div
      class="currency-input-balance"
    >
      Balance:
       
      $1,655.67
    </div>
  </div>
</div>
`;
